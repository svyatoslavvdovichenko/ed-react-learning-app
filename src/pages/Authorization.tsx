import { useHistory, useLocation } from 'react-router-dom'
import { Row, Tabs, Col, Typography } from 'antd'
import { Layout } from '../components/Layout'
import { SignIn } from '../components/onboarding/SignIn'
import { SignUp } from '../components/onboarding/SignUp'

const { TabPane } = Tabs
const { Title, Text } = Typography

export enum TabKeys {
  SignIn = 'sign-in',
  SignUp = 'sign-up',
}

export const Authorization = () => {
  const history = useHistory()
  const location = useLocation()

  let searchParams = new URLSearchParams(location.search)
  const activeTabKey = searchParams.get('tab')

  return (
    <Layout shouldBeCentered={true}>
      <Row justify="center" align="middle" style={{ height: '100%' }}>
        <Col>
          <Title style={{ marginBottom: 12, textAlign: 'center' }}>
            Exceed Challenger
          </Title>

          <Col
            style={{
              marginBottom: '26px',
              textAlign: 'center',
              whiteSpace: 'nowrap',
            }}
          >
            <Text type="secondary">
              Пожалуйста, выполните вход или регистрацию в системе
            </Text>
          </Col>

          <Tabs
            onChange={(key) =>
              history.push({
                pathname: location.pathname,
                search: `?tab=${key}`,
              })
            }
            activeKey={activeTabKey ? String(activeTabKey) : TabKeys.SignIn}
          >
            <TabPane key={TabKeys.SignIn} tab="Вход">
              <SignIn />
            </TabPane>

            <TabPane key={TabKeys.SignUp} tab="Регистрация">
              <SignUp />
            </TabPane>
          </Tabs>
        </Col>
      </Row>
    </Layout>
  )
}
