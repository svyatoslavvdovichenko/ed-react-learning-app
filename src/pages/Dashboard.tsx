import { FC } from 'react'
import { Row, Col, Button } from 'antd'
import { FiltersCard } from '../components/dashboard/FiltersCard'
import { TasksCard } from '../components/dashboard/TasksCard'
import { TaskSearchForm } from '../components/dashboard/TaskSearchForm'
import { Layout } from '../components/Layout'
import { AdminRequired } from '../hocs/AdminRequired'
import { useNavigate } from 'react-router-dom'
import { useQueryRequest } from '../hooks/useQueryRequest'
import { ISpecialization, ITechnology } from '../types'
import { useActions } from '../hooks/useActions'

export const Dashboard: FC = () => {
  const navigate = useNavigate()
  const { setReferences } = useActions()

  const { data: dataTech } = useQueryRequest<ITechnology[]>('v1/technologies/')
  const { data: dataSpec } = useQueryRequest<ISpecialization[]>(
    'v1/specializations/',
  )

  setReferences({ technologies: dataTech!, specializations: dataSpec! })

  return (
    <Layout>
      <Row justify="center">
        <Col flex="1170px">
          <Row justify="space-between" wrap={false}>
            <Col md={9} lg={7} xl={6}>
              <FiltersCard />
            </Col>

            <Col style={{ paddingLeft: 10 }} md={15} lg={17} xl={18}>
              <Row style={{ marginBottom: 30 }} wrap={false}>
                <Col flex="auto">
                  <TaskSearchForm />
                </Col>

                <AdminRequired>
                  <Col flex="100px" style={{ padding: '0 0 0 8px' }}>
                    <Button
                      onClick={() => navigate('/new-task')}
                      size="large"
                      block
                    >
                      Создать
                    </Button>
                  </Col>
                </AdminRequired>
              </Row>

              <TasksCard />
            </Col>
          </Row>
        </Col>
      </Row>
    </Layout>
  )
}
