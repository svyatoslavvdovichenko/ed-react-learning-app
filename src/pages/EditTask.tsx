import { Col, Row } from 'antd/lib/grid'
import { useHistory, useParams } from 'react-router-dom'
import { Loader } from '../components/common/Loader'
import { Layout } from '../components/Layout'
import { TaskForm } from '../components/task/TaskForm'
import { ERROR_NOTIFIFCATION_MESSAGE } from '../constants'
import { useApi } from '../hooks/useApi'
import { useQueryRequest } from '../hooks/useQueryRequest'
import { useTypedSelector } from '../hooks/useTypeSelector'

import { ITask } from '../types'
import {
  convertSpecializationOptions,
  convertTechnologiesOptions,
} from '../utils/converters'
import {
  sendErrorNotification,
  sendSuccessNotification,
} from '../utils/systemNotification'

export const EditTask = () => {
  const api = useApi()
  const navigate = useHistory()
  const { taskId } = useParams<{ taskId: string }>()

  const { data: task, isLoading } = useQueryRequest<ITask>(
    `v1/tasks/${taskId}/`,
  )

  const { technologies, specializations } = useTypedSelector(
    (state) => state.referenceReducer,
  )

  const handleUpdateTask = (values: any) => {
    api
      .put(`v1/tasks/${taskId}/`, {
        ...values,
        specialization: convertSpecializationOptions(
          values.specialization,
          specializations,
        ),
        technologies: convertTechnologiesOptions(
          values.technologies,
          technologies,
        ),
        attachments: values.attachments.map((item: string) => ({ url: item })),
      })
      .then(() => {
        sendSuccessNotification('Задание успешно обновлено')
        navigate.push('/dashboard')
      })
      .catch((err) => sendErrorNotification(ERROR_NOTIFIFCATION_MESSAGE))
  }

  return (
    <Layout>
      <Row justify="center">
        <Col flex="0 1 970px">
          {isLoading && <Loader />}

          {!isLoading && task && (
            <TaskForm
              isEdit
              initialValues={{
                title: task.title,
                specialization: task.specialization.title,
                description: task.description,
                technologies: task.technologies.map((item) => item.title),
                attachments: task.attachments.map((item) => item.url),
              }}
              handleSubmit={handleUpdateTask}
              submitButtonText="Изменить"
            />
          )}
        </Col>
      </Row>
    </Layout>
  )
}
