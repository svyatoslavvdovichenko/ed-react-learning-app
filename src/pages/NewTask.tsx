import { Col, Row } from 'antd/lib/grid'
import { useHistory } from 'react-router-dom'
import { Layout } from '../components/Layout'
import { TaskForm } from '../components/task/TaskForm'
import { ERROR_NOTIFIFCATION_MESSAGE } from '../constants'
import { useApi } from '../hooks/useApi'
import { useTypedSelector } from '../hooks/useTypedSelector'
import {
  convertTechnologiesOptions,
  convertSpecializationOptions,
} from '../utils/converters'
import {
  sendErrorNotification,
  sendSuccessNotification,
} from '../utils/systemNotification'

export const NewTask = () => {
  const api = useApi()
  const navigate = useHistory()

  const { technologies, specializations } = useTypedSelector(
    (state) => state.referenceReducer,
  )

  const handleCreateTask = (values: any) => {
    api
      .post('v1/tasks/', {
        ...values,
        technologies: convertTechnologiesOptions(
          values.technologies,
          technologies,
        ),
        specialization: convertSpecializationOptions(
          values.specialization,
          specializations,
        ),
        attachments: values.attachments.map((item: string) => ({ url: item })),
      })
      .then(() => {
        sendSuccessNotification('Задание успешно создано')
        navigate.push('/dashboard')
      })
      .catch((err) => sendErrorNotification(ERROR_NOTIFIFCATION_MESSAGE))
  }

  return (
    <Layout>
      <Row justify="center">
        <Col flex="0 1 970px">
          <TaskForm
            initialValues={{
              title: '',
              specialization: '',
              description: '',
              technologies: [],
              attachments: [],
            }}
            handleSubmit={handleCreateTask}
            submitButtonText="Добавить"
          />
        </Col>
      </Row>
    </Layout>
  )
}
