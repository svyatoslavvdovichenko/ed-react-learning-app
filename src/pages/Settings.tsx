import { Layout } from '../components/Layout'

export const Settings = () => {

  return <Layout>Settings</Layout>
}
