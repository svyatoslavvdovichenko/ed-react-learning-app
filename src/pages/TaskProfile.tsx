import { Button, Card, Col, Divider, Popconfirm, Row, Tag } from 'antd'
import { FC } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import { Breadcrumbs } from '../components/common/Breadcrumbs'
import { Loader } from '../components/common/Loader'
import { Layout } from '../components/Layout'
import { StyledText } from '../components/common/StyledComponents'
import { ITask } from '../types'
import { useQueryRequest } from '../hooks/useQueryRequest'
import { StatusButton } from '../components/task/StatusButton'
import ReactMarkdown from 'react-markdown'
import { useApi } from '../hooks/useApi'
import {
  sendErrorNotification,
  sendSuccessNotification,
} from '../utils/systemNotification'
import { ERROR_NOTIFIFCATION_MESSAGE } from '../constants'
import { AdminRequired } from '../hocs/AdminRequired'
import { DeleteOutlined, EditOutlined, LeftOutlined } from '@ant-design/icons'

export const TaskProfile: FC = () => {
  const { taskId } = useParams<{ taskId: string }>()
  const taskUrl = `v1/tasks/${taskId}/`

  const api = useApi()
  const navigate = useHistory()

  const { data: task, isLoading } = useQueryRequest<ITask>(taskUrl)

  const handleDeleteTask = () => {
    api
      .delete(taskUrl)
      .then(() => {
        sendSuccessNotification('Задание успешно удалено')
        navigate.push('/dashboard')
      })
      .catch((err) => sendErrorNotification(ERROR_NOTIFIFCATION_MESSAGE))
  }

  
  return (
    <Layout>
      <Breadcrumbs
        items={[
          {
            key: 'currentTask',
            title: task?.title ?? '-',
            link: `/task/${taskId}`,
          },
        ]}
      />
      {isLoading && <Loader fullScreen />}

      {!isLoading && task && (
        <Row justify="center">
          <Col flex="0 1 970px">
            <Card
              headStyle={{ padding: '16px 32px 0', maxWidth: '970px'}}
              bodyStyle={{ padding: '16px 32px 32px', maxWidth: '970px' }}
              title={task.title.length < 30 ? task.title : task.title.substr(0, 30) + "..."}
              extra={
                <AdminRequired>
                  <Popconfirm
                    title="Вы уверены, что хотите удалить эту задачу?"
                    onConfirm={handleDeleteTask}
                    okText="Да"
                    cancelText="Нет"
                  >
                    <Button
                      type="primary"
                      danger
                      style={{ margin: '0 8px 0 20px' }}
                    >
                      <DeleteOutlined />
                      Удалить задание
                    </Button>
                  </Popconfirm>

                  <Button
                    type="primary"
                    ghost
                    onClick={() => navigate.push(`/task/${taskId}/edit`)}
                  >
                    <EditOutlined />
                    Изменить
                  </Button>
                </AdminRequired>
              }
            >
              <Row align="middle" style={{ marginBottom: 20 }}>
                <StyledText>{task.specialization.title}</StyledText>

                <Divider
                  type="vertical"
                  style={{ background: '#555555', height: '20px' }}
                />

                {task.technologies.map((technology) => (
                  <Tag key={technology.id} color="green">
                    {technology.title}
                  </Tag>
                ))}
              </Row>

              <Row gutter={[0, 8]} style={{ marginBottom: 6 }}>
                <Col span={24}>
                  <StyledText>Описание задания</StyledText>
                </Col>

                <Col span={24} style={{ overflowWrap: "break-word" }}>
                  <ReactMarkdown>
                    {task.description}
                  </ReactMarkdown>
                </Col>
              </Row>

              {task.attachments.length > 0 && (
                <Row style={{ overflow: 'hidden' }}>
                  {task.attachments.map((attachment) => (
                    <Col
                      style={{ margin: '5px 0' }}
                      key={attachment.id}
                      span={24}
                    >
                      <a href={attachment.url} target="_blank">
                        {attachment.url}
                      </a>
                    </Col>
                  ))}
                </Row>
              )}

              <Row style={{ marginTop: 20 }} justify="end">
                <Button
                  type="primary"
                  ghost
                  onClick={() => navigate.push('/dashboard')}
                  style={{ height: 40, marginRight: 20 }}
                >
                  <LeftOutlined />
                  Назад
                </Button>

                <StatusButton />
              </Row>
            </Card>
          </Col>
        </Row>
      )}
    </Layout>
  )
}
