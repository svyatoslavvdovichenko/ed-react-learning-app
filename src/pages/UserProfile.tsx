import { Button, Card } from 'antd'
import { Col, Row } from 'antd/lib/grid'
import Text from 'antd/lib/typography/Text'
import { FC } from 'react'
import { Layout } from '../components/Layout'
import { StyledText } from '../components/common/StyledComponents'
import { useUser } from '../hooks/useUser'
import { TaskItem } from '../components/dashboard/TaskItem'
import { EditOutlined } from '@ant-design/icons'

export const UserProfile: FC = () => {
  const { user } = useUser()

  return (
    <Layout>
      <Row justify="center">
        <Col flex="0 1 970px">
          <Card>
            <Row justify="space-between">
              <Col span={8}>
                <StyledText>
                  {user?.first_name} {user?.last_name}
                </StyledText>
              </Col>
              
              <Col span={6}>
                <Button
                  type="primary"
                  ghost
                  style={{ height: 40, marginRight: 20 }}
                >
                  <EditOutlined />
                  Редактировать профиль
                </Button>
              </Col>
            </Row>

            <Row>
              <Text style={{ marginTop: '-10px' }} disabled>
                {user?.email}
              </Text>
            </Row>

            <Row style={{ marginTop: '30px' }}>
              <StyledText>Статистика пользователя</StyledText>
            </Row>

            <Row
              style={{
                width: '35%',
                display: 'flex',
                borderBottom: '1px solid #EEEEEE',
                flexWrap: 'nowrap',
              }}
            >
              <StyledText
                style={{
                  marginTop: '20px',
                  paddingBottom: '10px',
                  fontSize: '14px',
                  width: '100%',
                }}
              >
                Заданий выполнено
              </StyledText>

              <StyledText style={{ marginTop: '20px', fontSize: '14px' }}>
                {user?.completed_tasks.length}
              </StyledText>
            </Row>
          </Card>
        </Col>
      </Row>

      <Row justify="center" style={{ marginTop: '30px' }}>
        <Col flex="0 1 970px">
          <Card>
            <StyledText>Активные задания</StyledText>

            <Row>
              {!!user?.active_tasks.length ? (
                user?.active_tasks.map((task) => (
                  <TaskItem key={task.id} task={task} />
                ))
              ) : (
                <StyledText style={{ marginTop: '10px', fontSize: '14px' }}>
                  Нет активных заданий
                </StyledText>
              )}
            </Row>
          </Card>
        </Col>
      </Row>

      <Row justify="center" style={{ marginTop: '30px' }}>
        <Col flex="0 1 970px">
          <Card>
            <StyledText>Выполненные задания</StyledText>

            {!! user?.completed_tasks.length ? (
              user?.completed_tasks.map((task) => (
                <TaskItem key={task.id} task={task} />
              ))
            ) : (
              <p style={{ marginTop: '10px' }}>Нет выполненных заданий</p>
            )}
          </Card>
        </Col>
      </Row>
    </Layout>
  )
}
