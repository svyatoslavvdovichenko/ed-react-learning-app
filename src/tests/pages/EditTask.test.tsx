import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter, Route, Switch } from 'react-router-dom'
import { Dashboard } from '../../pages/Dashboard'
import { EditTask } from '../../pages/EditTask'

jest.mock('react-router-dom', () => {
  const originalModule = jest.requireActual('react-router-dom')

  return {
    ...originalModule,
    Link: () => <div />,
    useLocation: () => ({
      location: {
        search: 'search',
        pathname: 'pathname',
      },
    }),
    useParams: () => ({
      taskId: '1ffgj23dfaf234fg3',
    }),
  }
})

jest.mock('../../hooks/useQueryRequest', () => ({
  useQueryRequest: () => ({
    isLoading: false,
    data: {
      attachments: [{ id: 36, url: 'https://docs.google.com/document' }],
      description: 'Описание тестового задания',
      id: 23,
      specialization: {
        id: 1,
        title: 'frontend',
      },
      technologies: [
        {
          id: 123,
          title: 'react',
        },
      ],
      title: 'Название тестового задания',
    },
  }),
}))

const queryClient = new QueryClient()

const renderWithRouter = (component: JSX.Element) => {
  return {
    ...render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter initialEntries={['/task/:taskId/edit']}>
          <Switch>
            <Route path="/task/:taskId/edit">
              <EditTask />
            </Route>
            <Route path="/dashboard">
              <Dashboard />
            </Route>
          </Switch>
        </MemoryRouter>
      </QueryClientProvider>,
    ),
  }
}

jest.mock('../../hooks/useReferences', () => ({
  useReferences: () => ({
    technologies: [
      {
        id: 123,
        title: 'adaptive',
      },
    ],
    specialization: [
      {
        id: 1,
        title: 'frontend',
      },
    ],
  }),
}))

jest.mock('../../hooks/useUser', () => ({
  useUser: () => ({
    email: 'string',
    first_name: 'string',
    id: 213,
    is_admin: false,
    last_name: 'string',
    completed_tasks: [],
    active_tasks: [],
  }),
}))

jest.mock('../../hooks/useAuth', () => ({
  useAuth: () => ({
    onLogout: () => {},
    onLogin: () => {},
  }),
}))

describe('EditTask', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: () => {
        return {
          matches: false,
          addListener: () => {},
          removeListener: () => {},
        }
      },
    })
  })

  test('EditTask', () => {
    const { container } = render(<EditTask />)
    expect(container).toBeTruthy()
  })

  test('Placeholder works corrently', () => {
    render(<EditTask />)
    expect(screen.getByPlaceholderText(/Введите название/i)).toBeInTheDocument()
  })

  test('form works currently', () => {
    const { container, getByTestId, getByText, getByPlaceholderText } = render(<EditTask/>)
    const user = userEvent;
  
    user.type(getByPlaceholderText(/Введите название/i), 'John');
    user.type(getByPlaceholderText(/https:\/\/example\.com/i), 'Example text');
    user.type(getByPlaceholderText(/введите текст/i), "sqeqww");
    user.click(getByTestId(/specialization/i));
    user.click(getByText(/Frontend/i))
    user.click(getByTestId(/technologies/i));
    user.click(getByText(/React/i))
    user.click(getByText(/Добавить/i));

    expect(container).toMatchSnapshot();
  })
})

