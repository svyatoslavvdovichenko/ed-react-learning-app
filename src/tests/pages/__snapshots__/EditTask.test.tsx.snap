// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTask form works currently 1`] = `
<div>
  <section
    class="ant-layout"
    style="min-height: 100vh;"
  >
    <header
      class="ant-layout-header"
      style="background-color: white; border-bottom: 1px solid #F0F0F0;"
    >
      <div
        class="ant-row ant-row-space-between ant-row-middle"
        style="width: 100%; height: 100%;"
      >
        <div />
      </div>
    </header>
    <main
      class="ant-layout-content sc-ftTHYK kZHTsf"
    >
      <div
        class="ant-row ant-row-center"
      >
        <div
          class="ant-col"
          style="flex: 0 1 970px;"
        >
          <div
            class="ant-card ant-card-bordered"
          >
            <div
              class="ant-card-head"
              style="padding: 16px 32px 0px;"
            >
              <div
                class="ant-card-head-wrapper"
              >
                <div
                  class="ant-card-head-title"
                >
                  Редактирование задания
                </div>
              </div>
            </div>
            <div
              class="ant-card-body"
              style="padding: 16px 32px 32px;"
            >
              <form
                action="#"
              >
                <div
                  class="sc-dkrFOg jafsBz"
                >
                  <div
                    class="ant-row"
                    style="margin-bottom: 10px;"
                  >
                    Название задания
                  </div>
                  <input
                    class="ant-input sc-hLBbgP hibjcZ"
                    name="title"
                    placeholder="Введите название"
                    type="text"
                    value="Название тестового заданияJohn"
                  />
                </div>
                <div
                  class="ant-row"
                  style="margin-left: -4px; margin-right: -4px;"
                >
                  <div
                    class="ant-col ant-col-12"
                    style="padding-left: 4px; padding-right: 4px;"
                  >
                    <div
                      class="sc-dkrFOg jafsBz"
                    >
                      <div
                        class="ant-row"
                      >
                        <div
                          class="ant-row"
                          style="margin-bottom: 10px;"
                        >
                          Направление
                        </div>
                        <div
                          class="ant-select ant-select-single ant-select-show-arrow"
                          data-testid="specialization"
                        >
                          <div
                            class="ant-select-selector"
                          >
                            <span
                              class="ant-select-selection-search"
                            >
                              <input
                                aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                                aria-autocomplete="list"
                                aria-controls="rc_select_TEST_OR_SSR_list"
                                aria-expanded="false"
                                aria-haspopup="listbox"
                                aria-owns="rc_select_TEST_OR_SSR_list"
                                autocomplete="off"
                                class="ant-select-selection-search-input"
                                id="rc_select_TEST_OR_SSR"
                                readonly=""
                                role="combobox"
                                style="opacity: 0;"
                                type="search"
                                unselectable="on"
                                value=""
                              />
                            </span>
                            <span
                              class="ant-select-selection-item"
                              title="Frontend"
                            >
                              Frontend
                            </span>
                          </div>
                          <span
                            aria-hidden="true"
                            class="ant-select-arrow"
                            style="user-select: none;"
                            unselectable="on"
                          >
                            <span
                              aria-label="down"
                              class="anticon anticon-down ant-select-suffix"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                />
                              </svg>
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="ant-col ant-col-12"
                    style="padding-left: 4px; padding-right: 4px;"
                  >
                    <div
                      class="sc-dkrFOg jafsBz"
                    >
                      <div
                        class="ant-row"
                      >
                        <div
                          class="ant-row"
                          style="margin-bottom: 10px;"
                        >
                          Технологии
                        </div>
                        <div
                          class="ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search"
                          data-testid="technologies"
                        >
                          <div
                            class="ant-select-selector"
                          >
                            <div
                              class="ant-select-selection-overflow"
                            >
                              <div
                                class="ant-select-selection-overflow-item"
                                style="opacity: 1;"
                              >
                                <span>
                                  <span
                                    class="ant-tag ant-tag-green"
                                  >
                                    react
                                    <span
                                      aria-label="close"
                                      class="anticon anticon-close ant-tag-close-icon"
                                      role="img"
                                      tabindex="-1"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        data-icon="close"
                                        fill="currentColor"
                                        focusable="false"
                                        height="1em"
                                        viewBox="64 64 896 896"
                                        width="1em"
                                      >
                                        <path
                                          d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                        />
                                      </svg>
                                    </span>
                                  </span>
                                </span>
                              </div>
                              <div
                                class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                                style="opacity: 1;"
                              >
                                <div
                                  class="ant-select-selection-search"
                                  style="width: 0px;"
                                >
                                  <input
                                    aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                                    aria-autocomplete="list"
                                    aria-controls="rc_select_TEST_OR_SSR_list"
                                    aria-expanded="false"
                                    aria-haspopup="listbox"
                                    aria-owns="rc_select_TEST_OR_SSR_list"
                                    autocomplete="off"
                                    class="ant-select-selection-search-input"
                                    id="rc_select_TEST_OR_SSR"
                                    readonly=""
                                    role="combobox"
                                    style="opacity: 0;"
                                    type="search"
                                    unselectable="on"
                                    value=""
                                  />
                                  <span
                                    aria-hidden="true"
                                    class="ant-select-selection-search-mirror"
                                  >
                                    
                                     
                                  </span>
                                </div>
                              </div>
                            </div>
                          </div>
                          <span
                            aria-hidden="true"
                            class="ant-select-clear"
                            style="user-select: none;"
                            unselectable="on"
                          >
                            <span
                              aria-label="close-circle"
                              class="anticon anticon-close-circle"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="close-circle"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                                />
                              </svg>
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="sc-dkrFOg jafsBz"
                >
                  <div
                    class="ant-row"
                    style="margin-bottom: 10px;"
                  >
                    Текст задания
                  </div>
                  <textarea
                    class="ant-input"
                    data-testid="description"
                    name="description"
                    placeholder="Введите текст"
                  >
                    Описание тестового заданияsqeqww
                  </textarea>
                </div>
                <div
                  class="ant-row"
                  style="margin-left: -4px; margin-right: -4px; margin-bottom: 19px;"
                >
                  <div
                    class="ant-col ant-col-18"
                    style="padding-left: 4px; padding-right: 4px;"
                  >
                    <input
                      class="ant-input"
                      pattern="https://.*"
                      placeholder="https://example.com"
                      type="url"
                      value=""
                    />
                  </div>
                  <div
                    class="ant-col ant-col-6"
                    style="padding-left: 4px; padding-right: 4px;"
                  >
                    <button
                      ant-click-animating-without-extra-node="false"
                      class="ant-btn ant-btn-primary ant-btn-background-ghost ant-btn-block"
                      type="button"
                    >
                      <span
                        aria-label="link"
                        class="anticon anticon-link"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="link"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M574 665.4a8.03 8.03 0 00-11.3 0L446.5 781.6c-53.8 53.8-144.6 59.5-204 0-59.5-59.5-53.8-150.2 0-204l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3l-39.8-39.8a8.03 8.03 0 00-11.3 0L191.4 526.5c-84.6 84.6-84.6 221.5 0 306s221.5 84.6 306 0l116.2-116.2c3.1-3.1 3.1-8.2 0-11.3L574 665.4zm258.6-474c-84.6-84.6-221.5-84.6-306 0L410.3 307.6a8.03 8.03 0 000 11.3l39.7 39.7c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c53.8-53.8 144.6-59.5 204 0 59.5 59.5 53.8 150.2 0 204L665.3 562.6a8.03 8.03 0 000 11.3l39.8 39.8c3.1 3.1 8.2 3.1 11.3 0l116.2-116.2c84.5-84.6 84.5-221.5 0-306.1zM610.1 372.3a8.03 8.03 0 00-11.3 0L372.3 598.7a8.03 8.03 0 000 11.3l39.6 39.6c3.1 3.1 8.2 3.1 11.3 0l226.4-226.4c3.1-3.1 3.1-8.2 0-11.3l-39.5-39.6z"
                          />
                        </svg>
                      </span>
                      <span>
                        Добавить
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="ant-row"
                  style="margin-bottom: 16px; overflow: hidden;"
                >
                  <span
                    class="ant-tag ant-tag-processing"
                    style="margin: 3px 6px 3px 0px;"
                  >
                    https://docs.google.com/document
                    <span
                      aria-label="close"
                      class="anticon anticon-close sc-jrcTuL cFmhCp"
                      role="img"
                      tabindex="-1"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="close"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    class="ant-tag ant-tag-processing"
                    style="margin: 3px 6px 3px 0px;"
                  >
                    Exampletext
                    <span
                      aria-label="close"
                      class="anticon anticon-close sc-jrcTuL cFmhCp"
                      role="img"
                      tabindex="-1"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="close"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-row ant-row-end"
                >
                  <button
                    class="ant-btn ant-btn-primary ant-btn-background-ghost"
                    style="height: 40px; margin-right: 20px;"
                    type="button"
                  >
                    <span
                      aria-label="left"
                      class="anticon anticon-left"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="left"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                        />
                      </svg>
                    </span>
                    <span>
                      Назад
                    </span>
                  </button>
                  <button
                    class="ant-btn ant-btn-primary"
                    style="height: 40px;"
                    type="submit"
                  >
                    <span
                      aria-label="save"
                      class="anticon anticon-save"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="save"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M893.3 293.3L730.7 130.7c-7.5-7.5-16.7-13-26.7-16V112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V338.5c0-17-6.7-33.2-18.7-45.2zM384 184h256v104H384V184zm456 656H184V184h136v136c0 17.7 14.3 32 32 32h320c17.7 0 32-14.3 32-32V205.8l136 136V840zM512 442c-79.5 0-144 64.5-144 144s64.5 144 144 144 144-64.5 144-144-64.5-144-144-144zm0 224c-44.2 0-80-35.8-80-80s35.8-80 80-80 80 35.8 80 80-35.8 80-80 80z"
                        />
                      </svg>
                    </span>
                    <span>
                      Изменить
                    </span>
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="ant-layout-footer"
      style="background-color: rgb(240, 240, 240);"
    >
      Exceed Challenger 2022
    </footer>
  </section>
</div>
`;
